src/sagemaker_containers/transformers.py

```diff
@@ -0,0 +1,160 @@
+
+        Args:
+            model_dir (str): The directory where model files are stored.
+
+        Returns:
+            (obj) the loaded model.
+        """
+        raise NotImplementedError()
+
+    @staticmethod
+    def input_fn(input_data, content_type):
+        """Takes request data and de-serializes the data into an object for prediction.
+
+            When an InvokeEndpoint operation is made against an Endpoint running SageMaker model server,
+            the model server receives two pieces of information:
+
+                - The request Content-Type, for example "application/json"
+                - The request data content, which is at most 5 MB (5 * 1024 * 1024 bytes) in size.
+
+            The input_fn is responsible to take the request data and pre-process it before prediction.
+
+        Args:
+            input_data (obj): the request data content.
+            content_type (str): the request Content-Type.
+
+        Returns:
+            (obj): data ready for prediction.
+        """
+        return serializers.loads(input_data, content_type)
+
+    def predict_fn(self, model, data):
+        """Function responsible for model predictions.
+
+        Args:
+            model (obj): model loaded by model_fn
+            data: de-serializes data returned by input_fn
+
+         Returns:
+             (obj): data ready for prediction.
+        """
+        raise NotImplementedError()
+
+    @staticmethod
+    def output_fn(prediction, accept):
+        """Function responsible to serialize the prediction for the response.
+
+        Args:
+            prediction (obj): prediction returned by predict_fn .
+            accept (str): accept content-type expected by the client.
+
+        Returns:
+            (TransformSpec): a namedtuple with the following args:
+
+                * Args:
+                    serialized_prediction: the serialized data to return
+                    accept: the content-type that the data was transformed to.
+        """
+        serialized_prediction = serializers.dumps(prediction, accept)
+        return TransformSpec(serialized_prediction, accept)
+
+    def transform_fn(self, model, input_data, content_type, accept):
+        """Function responsible for input processing, prediction, and output processing.
+
+        Args:
+            model (obj): model loaded by model_fn
+            input_data (obj): the request data content.
+            content_type (str): the request Content-Type.
+            accept: the content-type that the data was transformed to.
+
+        Returns:
+
+        """
+        data = self.input_fn(input_data=input_data, content_type=content_type)
+        prediction = self.predict_fn(model=model, data=data)
+        return self.output_fn(prediction=prediction, accept=accept)
+
+    def initialize(self):
+        """function is called when the worker starts the Flask application.
```

Comment:
worker gets confusing here... i assume we are talking SM container worker class... but how does it related to gunicorn worker. especially important to be clear hear, since this function is all about startup order.

Reply:
also, they way i read this, it implies our default behavior is still to load model post-fork (once per worker). that's fine just asking.

Reply:
I rephrased to:
```
    def initialize(self):  # type: () -> None
        """Execute any initialization necessary to starting making predictions with the Transformer.

        The default implementation is used to load the model.

        This function is called by sagemaker_containers.worker.run, before starting the Flask application.
        The gunicorn server forks multiple workers, executing multiple Flask applications in parallel.
        This function will be called once per each worker.

        It does not have return type or arguments.
        """
```

Should we rename this function to prepare? Or something else?